File: 01-dsa-practices/code/primes/prime_numbers_sieve.go
package main

import (
	"fmt"
)

/**
 * Generate all prime numbers up to n using Sieve of Eratosthenes.
 */
func sieveOfEratosthenes(n int) []int {
	if n < 2 {
		return []int{}
	}

	isPrime := make([]bool, n+1)
	for i := range isPrime {
		isPrime[i] = true
	}

	isPrime[0] = false
	isPrime[1] = false

	for i := 2; i*i <= n; i++ {
		if isPrime[i] {
			for j := i * i; j <= n; j += i {
				isPrime[j] = false
			}
		}
	}

	primes := []int{}
	for i := 2; i <= n; i++ {
		if isPrime[i] {
			primes = append(primes, i)
		}
	}

	return primes
}

func main() {
	n := 30
	primes := sieveOfEratosthenes(n)
	fmt.Printf("Primes up to %d: %v\n", n, primes)
}

File: 01-dsa-practices/code/primes/prime_numbers_sieve.py
from typing import List

def sieve_of_eratosthenes(n: int) -> List[int]:
    """
    Generate all prime numbers up to n using Sieve of Eratosthenes.

    Args:
        n (int): Upper bound (inclusive)

    Returns:
        List[int]: List of prime numbers ≤ n
    """
    if n < 2:
        return []

    is_prime = [True] * (n + 1)
    is_prime[0] = is_prime[1] = False

    i = 2
    while i * i <= n:
        if is_prime[i]:
            for j in range(i * i, n + 1, i):
                is_prime[j] = False
        i += 1

    return [i for i in range(2, n + 1) if is_prime[i]]


if __name__ == "__main__":
    n = 30
    primes = sieve_of_eratosthenes(n)
    print(f"Primes up to {n}: {primes}")
